File: markovTweet.py
from collections import defaultdict
import random
import time
filename = "frank10.txt"

f = open(filename, 'r')

markovMatrix = defaultdict(dict)

starters = []
enders = []

for line in f:
    splitLine = line.replace(",","").split()
    i = 0
    while i < len(splitLine) - 1:

        #get 2 adjacent words
        first = splitLine[i]
        second = splitLine[i+1]
        #check for edge words
        if "." in first:
            starters += [second]
            enders += [first]
            i += 1
            continue
        elif "." in second:
            enders += [second]

        #add to adjacent word matrix
        if first in markovMatrix.keys():
            if second in markovMatrix.get(first):
                markovMatrix[first][second] += 1
            else:
                markovMatrix[first][second] = 1
        else:
            markovMatrix[first][second] = 1
        i += 1

#given a 1D dictionary, pick one key weighted random based on the values
def weightedChoice(dict):
    pool = []
    for key in dict.keys():
        for i in range(0,dict[key]):
            pool += [key]
    return random.choice(pool)

def makeTweet():
    tweet = [random.choice(starters)]
    for i in range(0,280):

        if markovMatrix.get(tweet[i]) == None:
            tweet += [random.choice(enders)]
            break
        nextWord = weightedChoice(markovMatrix.get(tweet[i]))
        tweet += [nextWord]
        if nextWord in enders:
            break
    print " ".join(tweet)


for i in range(0,3):
    makeTweet()
    print "\n"

File: markovTweet.cpp
#include <sstream>
#include <fstream>
#include <iostream>
#include <string>
#include <vector>
#include <map>
using namespace std;

vector<string> starters; 
vector<string> enders; 

std::map<string, map<string,int> > markovMatrix;


void collectSampleData(string fileName)
{
	ifstream file;
	file.open(fileName.c_str());
	string word;
	string firstWord,secondWord;
	char wordEnding;
	int i = 0;
	int endOfFile = 0;
	while(!endOfFile) 
	{ /*
		printf("%s\n",file.get());
		if (i == 0) 
		{
			firstWord = word;
			i = 1;
		} 
		else 
		{
			secondWord = word;
			i = 0;
		}
		*/
		
		
		file >> firstWord;
		file >> secondWord;
		
		
		char firstWordEnding = firstWord[firstWord.size() -1];
		char secondWordEnding = secondWord[secondWord.size() -1];
		if(firstWordEnding == '.' || firstWordEnding == ',' || firstWordEnding == '?' || firstWordEnding == '!' ||
           firstWordEnding == ')' || firstWordEnding == ':' || firstWordEnding == ';' || firstWordEnding == '"')
        {
			starters.push_back(secondWord);
			enders.push_back(firstWord);
		}
		else
		{
			if (secondWordEnding == '.' || secondWordEnding == ',' || secondWordEnding == '?' || secondWordEnding == '!' ||
           secondWordEnding == ')' || secondWordEnding == ':' || secondWordEnding == ';' || secondWordEnding == '"')
		   {
			   enders.push_back(secondWord);
		   }
		}
		
		//if first in markovMatrix.keys();
		printf("checking for %s\n",firstWord.c_str());
		if (markovMatrix.find(firstWord) != markovMatrix.end())
		{
			printf("checking for %s\n",secondWord.c_str());
			if (markovMatrix.find(firstWord)->second.find(secondWord)->first != "")
			{
				printf("No second. inserting %s\n",secondWord.c_str());
				markovMatrix[firstWord][secondWord] += 1;
			} 
			else
			{
				markovMatrix[firstWord][secondWord] = 1;
			}
		}
		else
		{
			printf("No first. Inserting pair %s\n",firstWord.c_str());
			markovMatrix[firstWord][secondWord] = 1;
		}
		
	}
	
}

int main () {
	collectSampleData("frank10.txt");
}
